server/service/kv/dao.go

```diff
@@ -0,0 +1,225 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kvsvc
+
+import (
+	"context"
+	"fmt"
+	"github.com/apache/servicecomb-kie/pkg/model"
+	"github.com/apache/servicecomb-kie/server/db"
+	"github.com/apache/servicecomb-kie/server/service/history"
+	"github.com/apache/servicecomb-kie/server/service/label"
+	"github.com/go-mesh/openlogging"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/bson/primitive"
+	"go.mongodb.org/mongo-driver/mongo"
+)
+
+//createKey get latest revision from history
+//and increase revision of label
+//and insert key
+func createKey(ctx context.Context, kv *model.KVDoc) (*model.KVDoc, error) {
+	c, err := db.GetClient()
+	if err != nil {
+		return nil, err
+	}
+	r, err := label.GetLatestLabel(ctx, kv.LabelID)
+	if err != nil {
+		if err != db.ErrRevisionNotExist {
+			openlogging.Error(fmt.Sprintf("get latest [%s][%s] in [%s],err: %s",
+				kv.Key, kv.Labels, kv.Domain, err.Error()))
+			return nil, err
+		}
+		//the first time labels is created, at this time, labels has no revision yet
+		//after first key created, labels got revision 1
+		r = &model.LabelRevisionDoc{Revision: 0}
+	}
+	if r != nil {
+		r.Revision = r.Revision + 1
+	}
+	collection := c.Database(db.Name).Collection(db.CollectionKV)
+	res, err := collection.InsertOne(ctx, kv)
+	if err != nil {
+		return nil, err
+	}
+	objectID, _ := res.InsertedID.(primitive.ObjectID)
+	kv.ID = objectID
+	kvs, err := findKeys(ctx, bson.M{"label_id": kv.LabelID}, true)
+	//Key may be empty When delete
+	if err != nil && err != db.ErrKeyNotExists {
+		return nil, err
+	}
+	revision, err := history.GetAndAddHistory(ctx, kv.LabelID, kv.Labels, kvs, kv.Domain)
+	if err != nil {
+		openlogging.Warn(
+			fmt.Sprintf("can not updateKey version for [%s] [%s] in [%s]",
+				kv.Key, kv.Labels, kv.Domain))
+	}
+	openlogging.Debug(fmt.Sprintf("create %s with labels %s value [%s]", kv.Key, kv.Labels, kv.Value))
+	kv.Revision = revision
+	return kv, nil
+
+}
+
+//updateKey get latest revision from history
+//and increase revision of label
+//and updateKey and them add new revision
+func updateKey(ctx context.Context, kv *model.KVDoc) (int, error) {
```

Comment:
I prefer to change it to updateKV, otherwise, I thought it was just updating the key